x-pack/plugins/triggers_actions_ui/public/application/sections/rules_list/components/rules_list_column_selector.tsx

```diff
@@ -0,0 +1,133 @@
+  EuiTableComputedColumnType,
+  EuiTableFieldDataColumnType,
+  useDataGridColumnSelector,
+} from '@elastic/eui';
+import React, { useMemo, useState } from 'react';
+import { useKibana } from '../../../../common/lib/kibana';
+import { RuleTableItem } from '../../../../types';
+
+type RulesListTableColumns =
+  | EuiTableFieldDataColumnType<RuleTableItem>
+  | EuiTableComputedColumnType<RuleTableItem>
+  | EuiTableActionsColumnType<RuleTableItem>;
+
+export type RulesListColumns = {
+  id?: RulesListVisibleColumns;
+  selectorName?: string;
+} & RulesListTableColumns;
+
+export type RulesListVisibleColumns =
+  | 'ruleName'
+  | 'ruleTags'
+  | 'ruleExecutionStatusLastDate'
+  | 'ruleSnoozeNotify'
+  | 'ruleScheduleInterval'
+  | 'ruleExecutionStatusLastDuration'
+  | 'ruleExecutionPercentile'
+  | 'ruleExecutionSuccessRatio'
+  | 'ruleExecutionStatus'
+  | 'ruleExecutionState';
+
+const OriginalRulesListVisibleColumns: RulesListVisibleColumns[] = [
+  'ruleName',
+  'ruleTags',
+  'ruleExecutionStatusLastDate',
+  'ruleSnoozeNotify',
+  'ruleScheduleInterval',
+  'ruleExecutionStatusLastDuration',
+  'ruleExecutionPercentile',
+  'ruleExecutionSuccessRatio',
+  'ruleExecutionStatus',
+  'ruleExecutionState',
+];
+
+interface RulesListColumnSelector {
+  allRuleColumns: RulesListColumns[];
+  rulesListKey?: string;
+  visibleColumns?: RulesListVisibleColumns[];
+}
+
+type UseRulesListColumnSelector = [RulesListTableColumns[], React.ReactNode];
+
+const RULES_LIST_COLUMNS_KEY = 'triggersActionsUi_rulesListColumns';
+
+export const useRulesListColumnSelector = ({
+  allRuleColumns,
+  rulesListKey = RULES_LIST_COLUMNS_KEY,
+  visibleColumns = OriginalRulesListVisibleColumns,
+}: RulesListColumnSelector): UseRulesListColumnSelector => {
+  const { storage } = useKibana().services;
+  const storageVisibleColumns = storage?.get(rulesListKey);
+  const [localVisibleColumns, setLocalVisibleColumns] = useState<string[]>(
+    storageVisibleColumns ?? visibleColumns
+  );
+
+  const rulesColumns = useMemo(() => {
+    const columnsToAddAtTheEnd = allRuleColumns.filter((col) => col.id === undefined);
+    const columns = localVisibleColumns.map((coldId) => {
+      const { id, selectorName, ...colAttr } = allRuleColumns.find((col) => col.id === coldId) ?? {
+        id: '',
+        selectorName: '',
+      };
+      return colAttr as RulesListTableColumns;
+    });
+    columns.push(...columnsToAddAtTheEnd);
+    return columns;
+  }, [allRuleColumns, localVisibleColumns]);
+
+  const rulesListColumnSelector = useMemo(
+    () =>
+      allRuleColumns
+        .filter((col) => col.id != null)
+        .map((col) => ({ id: col.id })) as EuiDataGridColumn[],
+    [allRuleColumns]
+  );
+
+  const rulesListColumnVisibility = useMemo(
+    () => ({
+      visibleColumns: localVisibleColumns,
+      setVisibleColumns: (col: string[]) => {
+        if (rulesListKey) {
+          storage?.set(rulesListKey, col);
+        }
+        setLocalVisibleColumns(col);
```

Comment:
nitpick: might be better to put this in a `useEffect` so it's easier to extend this component later without worrying about syncing the localstorage and local state. 

```
useEffect(() => {
  if (rulesListKey) {
    storage?.set(rulesListKey, col);
  }
}, [rulesListKey, col]);
```

Reply:
I wanted to avoid an useEffect, I always prefer to add the code where it needs to be, I do not like to rely on the dependency of the useEffect. Like that I can avoid logic like it is at initialization, or check if the col are equal or not.

Reply:
yep good point! useEffect does invite a lot of misuse.

Reply:
instead of using useEffect, we just create a specific function so it is more readable to the user of what's going on